### PR TITLE
Add a CI for contracts package and frontend

### PR DESCRIPTION
**Pull request by gurrpi**
This PR activates CI to check pull requests for contracts package or frontend.
Documentation for CI/CD pipeline can be found [here](https://hackmd.io/@mdhackmdkakao/rkoVmtr75).

### Every PR will trigger tests on

contracts package

unit test and linting
qv test
qf test

### frontend

Check build of packages/app, packages/hooks, packages/subgraph
Lighthouse audit (only if there are changes to the above packages)

### Push event on main branch

Code coverage report will be generated for push events for the main branch.

### CODE REVIEW

File: .github/workflows/contracts-build.yml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  cwd: ${{github.workspace}}/packages/contracts
+
+concurrency: 
+  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
+  cancel-in-progress: true
+
+jobs:
+
+  unit-test-and-lint:
+    runs-on: ubuntu-20.04
```
Comment (daodesigner) [inline .github/workflows/contracts-build.yml:19]: need ^21.04

File: .github/workflows/contracts-coverage.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+          yarn workspace @quadratic-funding/contracts run lint:solidity
+          yarn workspace @quadratic-funding/contracts run lint:js
+
+      - name: Test Contract
+        run: |
+          yarn workspace @quadratic-funding/contracts run generate:abi
+          yarn workspace @quadratic-funding/contracts run test:coverage
+
+      - name: Upload coverage report
+        uses: codecov/codecov-action@v2
```
Comment (daodesigner) [inline .github/workflows/contracts-coverage.yml:41]: does this cost $$?

File: .github/workflows/frontend-build.yml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        uses: actions/setup-node@v2
+        with:
+          node-version: 16
+
+      - name: Install dependencies
+        run: yarn install --frozen-lockfile
+          
+      - name: Build
+        run: |
+          CI=false yarn workspace @qfi/app build
```
Comment (daodesigner) [inline .github/workflows/frontend-build.yml:32]: build hooks > ui > app

File: .github/workflows/frontend-lighthouse.yml [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+on: 
+  issue_comment:
+    types: [ edited ]
+
+jobs:
+  capture_vercel_preview_url:
+    name: Capture Vercel preview URL
+    runs-on: ubuntu-20.04
+    steps:
+      - uses: aaimio/vercel-preview-url-action@v1
```
Comment (daodesigner) [inline .github/workflows/frontend-lighthouse.yml:12]: preffer web3 equivalent: https://fleek.co/

File: .github/workflows/frontend-lighthouse.yml [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        uses: actions/github-script@v3
+        with:
+          github-token: ${{secrets.GITHUB_TOKEN}}
+          script: |
+            const result = ${{ steps.lighthouse_audit.outputs.manifest }}[0].summary
+            const links = ${{ steps.lighthouse_audit.outputs.links }}
+            const formatResult = (res) => Math.round((res * 100))
+            Object.keys(result).forEach(key => result[key] = formatResult(result[key]))
+            const score = res => res >= 90 ? '🟢' : res >= 50 ? '🟠' : '🔴'
+            const comment = [
```
Comment (daodesigner) [inline .github/workflows/frontend-lighthouse.yml:37]: <3

**Review by daodesigner — CHANGES_REQUESTED** [review]
good job so far, just a few nits

**Comment by daodesigner** [comment]
also, seems like checks for PR are failing so glad that's working too 🎉

File: .github/workflows/contracts-coverage.yml (hunk 2 above)
Comment (gurrpi) [reply .github/workflows/contracts-coverage.yml:41]: It's entirely free for open source project.

File: .github/workflows/contracts-build.yml (hunk 1 above)
Comment (gurrpi) [reply .github/workflows/contracts-build.yml:19]: Can you elaborate more? Test suite of contract package runs well on any linux distribution and macOS.

Comment (daodesigner) [reply .github/workflows/contracts-build.yml:19]: right now we're mocking out some of the steps, once proofs are actually being generated we need 21.04 distro to run the binaries that generate proofs

File: .github/workflows/frontend-lighthouse.yml (hunk 4 above)
Comment (daodesigner) [reply .github/workflows/frontend-lighthouse.yml:12]: ok, check in with Barry/Thore if any of these services cost $$

Comment (gurrpi) [reply .github/workflows/frontend-lighthouse.yml:12]: Ack, I think we can rely on free tier plan at the moment. And also I left feature request to Fleek team. Hope we could migrate to it near future.

**Comment by gurrpi** [comment]
@ChallaHalla Could you please fix linter errors for unit tests? You can see errors from [build status](https://github.com/quadratic-funding/qfi/runs/5853249496?check_suite_focus=true).

fyi: Workaround that automatically fix lint error will be merged soon, it won't be long before automation will handle consistent code style.

File: .github/workflows/contracts-build.yml (hunk 1 above)
Comment (gurrpi) [reply .github/workflows/contracts-build.yml:19]: Ack, I will setup our own self-hosted runner (CI instance) on AWS and will make integration tests run on that instance; GitHub-hosted runner didn't support 21.04.

Comment (daodesigner) [reply .github/workflows/contracts-build.yml:19]: Ok let's just open this up as a new issue

**Comment by daodesigner** [comment]
Opening a new issue for linter failing and merging this.

**Pull request merged**